src/db.c

```diff
@@ -1604,25 +1604,143 @@ int *getKeysPrepareResult(getKeysResult *result, int numkeys) {
+                } else {
+                    serverAssert(spec->fk.range.lastkey == -1);
+                    last = first + ((argc-first)/spec->fk.range.limit + spec->fk.range.lastkey);
+                }
+            }
+        } else {
+            step = spec->fk.keynum.keystep;
+            long long numkeys;
+
+            if (!string2ll(argv[first+spec->fk.keynum.keynumidx]->ptr,sdslen(argv[first+spec->fk.keynum.keynumidx]->ptr),&numkeys)) {
+                result->numkeys = 0;
+                return 0;
+            }
+
+            first += spec->fk.keynum.firstkey;
+            last = first + (int)numkeys-1;
+        }
+
+        int count = ((last - first)+1);
+        keys = getKeysPrepareResult(result, count);
+
+        for (i = first; i <= last; i += step) {
+            if (i >= argc) {
+                /* Modules commands, and standard commands with a not fixed number
+                 * of arguments (negative arity parameter) do not have dispatch
+                 * time arity checks, so we need to handle the case where the user
+                 * passed an invalid number of arguments here. In this case we
+                 * return no keys and expect the command implementation to report
+                 * an arity or syntax error. */
+                if (cmd->flags & CMD_MODULE || cmd->arity < 0) {
+                    result->numkeys = 0;
+                    return 0;
+                } else {
+                    serverPanic("Redis built-in command declared keys positions not matching the arity requirements.");
+                }
+            }
+            keys[k].pos = i;
+            keys[k++].flags = spec->flags;
+        }
+    }
+
+    result->numkeys = k;
+    return k;
+}
+
+/* Return all the arguments that are keys in the command passed via argc / argv. 
+ * This function will eventually replace getKeysFromCommand.
+ *
+ * The command returns the positions of all the key arguments inside the array,
+ * so the actual return value is a heap allocated array of integers. The
+ * length of the array is returned by reference into *numkeys.
+ * 
+ * Along with the position, this command also returns the flags that are
+ * associated with how Redis will access the key.
+ *
+ * 'cmd' must be point to the corresponding entry into the redisCommand
+ * table, according to the command name in argv[0].
+ *
+ * This function uses the command table if a command-specific helper function
+ * is not required, otherwise it calls the command-specific function. */
+int getKeysFromCommandWithSpecs(struct redisCommand *cmd, robj **argv, int argc, getKeysResult *result) {
+    if (cmd->flags & CMD_MODULE_GETKEYS) {
+        return moduleGetCommandKeysViaAPI(cmd,argv,argc,result);
+    } else {
+        int ret = getKeysUsingKeySpecs(cmd,argv,argc,result);
+        if (ret) {
+            return ret;
+        } else if (!(cmd->flags & CMD_MODULE) && cmd->getkeys_proc) {
+            return cmd->getkeys_proc(cmd,argv,argc,result);
```

Comment:
Now that I understand this better, I believe we want to trust the specs over the procs here. I will omit adding the flags.

Reply:
there are some cases where the specs are limited (can't find all keys).
specifically MIGRATE, when the the AUTH password is "KEYS".
since this is for ACL, maybe we can't afford to make any compromises here?
maybe we wanna use the `incomplete` flag, to decide between these approaches?
i.e. always prefer the key-specs, unless the `incomplete` flag is present, in which case we make the `getkeys_proc` mandatory.
@guybe7 WDYT?

Reply:
yes, prefer key-specs, unless one of them has "incomplete"